Add pair style lj/pirani

mateors6 opened the pull request:
**Summary**

Rename of a previous fork and pull request with the required modifications submitted. 

This pull adds a new pair style: an Improved Lennard-Jones (ILJ) potential, by Pirani et al. Phys. Chem. Chem. Phys., 2008,10, 5489-5503


**Related Issue(s)**


**Author(s)**

Mateo Rodríguez (mateorsuarez@gmail.com) ( Molecular Interactions Group INTERMOL IFF-CSIC, http://intermol.iff.csic.es)

Optimization of the code by Axel Kohlmeyer (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**


**Implementation Notes**


**Post Submission Checklist**

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

mateors6 committed:
Add files via upload

mateors6 committed:
Add files via upload

mateors6 committed:
Delete doc/pair_ilj_cut.rst

mateors6 committed:
Add files via upload

mateors6 committed:
Merge branch 'develop' into ILJ/CUT

mateors6 committed:
Delete doc/pair_ilj_cut.html

mateors6 committed:
Update Commands_pair.rst

mateors6 committed:
Update pair_style.rst

mateors6 committed:
Merge branch 'lammps:develop' into lj/improved/cut

mateors6 committed:
Update .gitignore

mateors6 committed:
Delete src/EXTRA-PAIR/pair_ilj_cut.cpp

mateors6 committed:
Delete src/EXTRA-PAIR/pair_ilj_cut.h

mateors6 committed:
Add files via upload

mateors6 committed:
Update and rename pair_ilj_cut.rst to pair_lj_improved_cut.rst

mateors6 committed:
Update pair_lj_improved_cut.rst

mateors6 committed:
Update .gitignore

akohlmey committed:
add force style test

akohlmey committed:
replace slow pow(x,2.0) with square(x) and cache repeatedly used pow() calls

akohlmey committed:
address multiple issues with formatting, references, non-ASCII characters, spelling

mateors6 committed:
Update pair_lj_improved_cut.cpp

akohlmey committed:
add another small optimization

mateors6 committed:
Update pair_lj_improved_cut.cpp

Corrected energy calculation for r-RESPA

mateors6 committed:
Update Commands_pair.rst

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.rst to pair_lj_pirani.rst

New name

mateors6 committed:
Update pair_style.rst

New name

mateors6 committed:
Update .gitignore

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.cpp to pair_lj_pirani.cpp

New name

mateors6 committed:
Update and rename pair_lj_improved_cut.h to pair_lj_pirani.h

New name

mateors6 committed:
Update and rename mol-pair-lj_improved_cut.yaml to mol-pair-lj_pirani.yaml

New name

akohlmey committed:
apply clang-format

megmcca reviewed (approved):
suggested name change from @sjplimp implemented! thanks @mateors6

akohlmey committed:
Add OPENMP package version of pair style lj/pirani

akohlmey committed:
Update extract(), document, and test

akohlmey commented:
@mateors6 I've added an OPENMP version and updated / documented / tested the extract() function

mateors6 commented:
> @mateors6 I've added an OPENMP version and updated / documented / tested the extract() function

That's great. Thank you! 

If I can (or I should) do anything else, please let me know.

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+
+
+An additional parameter, :math:`\alpha`, has been introduced in order
+to be able to recover the traditional Lennard-Jones (LJ) 12-6 with an adequate
```
an adequate --> a specific

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+it is straightforward to prove that LJ 12-6 is obtained.
+
+
+This potential provides some advantages with respect to the LJ
```
standard LJ

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+
+
+This potential provides some advantages with respect to the LJ
+potential and can be really useful for molecular dynamics simulations,
```
potential as explained in ref:Pirani ...  (drop really useful for MD)

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+as one can see from :ref:`(Pirani) <Pirani>`.
+It can be used for neutral-neutral (:math:`\gamma = 6`),
+ion-neutral (:math:`\gamma = 4`) or ion-ion systems (:math:`\gamma = 1`).
+It removes most of the issues at short- and long-range of the LJ model.
```
These settings remove issues at short- and long-range for these systems when a standard LJ model is used.

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+
+It is possible to verify that using (:math:`\alpha= 4`), (:math:`\beta= 6`)
+and (:math:`\gamma = 6`), at the equilibrium distance,
+the first and second derivatives of ILJ coincide with those of LJ 12-6
```
coincide with --> match

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+It is possible to verify that using (:math:`\alpha= 4`), (:math:`\beta= 6`)
+and (:math:`\gamma = 6`), at the equilibrium distance,
+the first and second derivatives of ILJ coincide with those of LJ 12-6
+( and the reduced force constant amounts to the typical 72).
```
don't understand this parenthesized statement - can you reword it ?

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+and (:math:`\gamma = 6`), at the equilibrium distance,
+the first and second derivatives of ILJ coincide with those of LJ 12-6
+( and the reduced force constant amounts to the typical 72).
+In this case, LJ provides a long-range coefficient with a factor of 2 compared
```
Awkward English - what is a long-range coefficient?  Do you mean the standard LJ energy (or force) is 2x stronger than ILJ at long distances?

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+the first and second derivatives of ILJ coincide with those of LJ 12-6
+( and the reduced force constant amounts to the typical 72).
+In this case, LJ provides a long-range coefficient with a factor of 2 compared
+with the ILJ. Also, the short-range interaction is overestimated by LJ.
```
strength of the short-range interaction

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+The ILJ potential solves both problems.
+
+
+The analysis of a diverse amount of systems verified that (:math:`\alpha= 4`)
```
As discussed in the Perani paper (give ref again), analyses of a variety of systems showed that alpha=4 generally works very well.

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+multiple charged ions) this factor may take a slightly different value.
+The parameter (:math:`\beta`) codifies the hardness (polarizability) of the
+interacting partners, and for neutral-neutral systems it ranges from 6 to 11.
+Moreover, the modulation of (:math:`\beta`) permits to indirectly consider the
```
Moreover, the modulation of (:math:`\beta`) can model additional
interaction effects, such as charge transfer in the
perturbative limit, and can thus mitigate the effect of some uncertainty in the data.  (What does "data" mean - do you mean uncertainty in other input parameters, e.g. alpha or gamma ?_

sjplimp reviewed (changes requested):
See my doc page comments to clarify details for prospective users of this new pair style

mateors6 committed:
Update pair_lj_pirani.rst

Corrections

mateors6 commented:
> See my doc page comments to clarify details for prospective users of this new pair style

Thank you for your review. I have applied the required corrections.

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,155 @@
+
+Pair style *lj/pirani* computes pairwise interactions from an Improved
+Lennard-Jones (ILJ) potential according to :ref:`(Pirani) <Pirani>`:
+
```
Here is where I think you need a sentence of two about what kinds of systems ILJ could be used for (compared to standard LJ).  Also for charged systems you need to explain if ILJ is used with a Coulomb potential (via pair hybrid), or ILJ effectively models LJ + Coulombic effects w/out needing to define/assign charges.  Then show the formula for ILJ.

sjplimp replied:
Also mention what physics the ILJ potential is attempting to capture - is it screening effects in charged systems?

mateors6 replied:
That's interesting. I will take a look at it and add that kind of information.

Nonetheless, in a previous version I had a couple of examples about recent uses of the ILJ potential:

- Model the interaction between a graphdiyne (GDY) membrane and molecular hydrogen.

(Phys. Chem. Chem. Phys., 2022,24, 15840-15850)

- Evaluate transport properties in gaseous mixtures.

(Physics of Fluids 1 August 2022; 34 (8): 087106)

I was asked to remove these lines. I understand this necessary piece of information should be something more general than these two uses, please correct me if I am wrong.

sjplimp replied:
If I was the one who asked you to remove than info previously, then my apologies.  It would be OK to cite these 2 papers with the brief description you list as example applications.  But a more general description of what systems ILJ is designed for, and answering the yes/no on Coulombic terms will also be useful context for users.

sjplimp reviewed (changes requested):
Made an additional comment at the top of the doc page - didn't see that you addressed this point from my earlier comments.

mateors6 committed:
Update pair_lj_pirani.rst

Documentation update

mateors6 commented:
> Made an additional comment at the top of the doc page - didn't see that you addressed this point from my earlier comments.

@sjplimp I have uploaded a new version of the file. The discussion of the potential should be clearer now.

sjplimp commented on doc/src/pair_lj_pirani.rst:
```diff
@@ -0,0 +1,163 @@
+This force field can be used for neutral-neutral (:math:`\gamma = 6`),
+ion-neutral (:math:`\gamma = 4`) or ion-ion systems (:math:`\gamma = 1`).
+Notice that this implementation does not include possible electrostatic
+interactions which should be eventually added by means of a hybrid style
```
I now see that you addressed Coulombic interaction here:

Please re-word to:
does not include explicit electrostatic interactions.  If these are desired, this pair style should be used along with a Coulomb pair style (put link to pair_coul.html) and optional long-range Coulombics, which can be done using the pair_style hybrid (put link) and kspace style (put link) commands.

sjplimp reviewed (commented):
last comment - after addressing these points (cites and Coulomb sentence), the PR should be ready to merge

mateors6 committed:
Update pair_lj_pirani.rst

mateors6 commented:
> last comment - after addressing these points (cites and Coulomb sentence), the PR should be ready to merge

@sjplimp Thank you so much for your comments and corrections. The Coulomb sentence has been rewritten. 

I have finally not added those two references. It was @akohlmey comment and he is probably right on saying that information is not necessary, since it can be obtained from the citations of the main paper.

By the way, I will probably try to add in the future another pair lj/pirani/coul so it also computes electrostatic interactions, I think that could be really useful. However, I still have to do my research on how LAMMPS treats partial charges of molecules.

Thank you both for your help!

akohlmey commented:
> I will probably try to add in the future another pair lj/pirani/coul so it also computes electrostatic interactions,

Please note, that there are multiple ways to include Coulomb interactions (plain cutoff, various smoothed/compensated cutoff methods, long-range solvers).

A simple way to include Coulomb is to use pair style hybrid/overlay and combine your pair style with coul/cut, coul/long, or any of the other Coulomb-only pair styles. Having the combined pair style is mostly for optimizing performance and as base pair style for accelerated versions.

mateors6 commented:
> > I will probably try to add in the future another pair lj/pirani/coul so it also computes electrostatic interactions,
> 
> Please note, that there are multiple ways to include Coulomb interactions (plain cutoff, various smoothed/compensated cutoff methods, long-range solvers).
> 
> A simple way to include Coulomb is to use pair style hybrid/overlay and combine your pair style with coul/cut, coul/long, or any of the other Coulomb-only pair styles. Having the combined pair style is mostly for optimizing performance and as base pair style for accelerated versions.

Thanks, I'll keep that in mind.
